code/datums/action.dm

```diff
@@ -0,0 +1,406 @@
+	name = "Toggle Helmet"
+
+/datum/action/item_action/toggle_jetpack
+	name = "Toggle Jetpack"
+
+/datum/action/item_action/jetpack_stabilization
+	name = "Toggle Jetpack Stabilization"
+
+/datum/action/item_action/jetpack_stabilization/IsAvailable()
+	var/obj/item/weapon/tank/jetpack/J = target
+	if(!istype(J) || !J.on)
+		return 0
+	return ..()
+
+/datum/action/item_action/hands_free
+	check_flags = AB_CHECK_CONSCIOUS
+
+/datum/action/item_action/hands_free/activate
+	name = "Activate"
+
+/datum/action/item_action/toggle_research_scanner
+	name = "Toggle Research Scanner"
+	button_icon_state = "scan_mode"
+
+/datum/action/item_action/toggle_research_scanner/Trigger()
+	if(IsAvailable())
+		owner.research_scanner = !owner.research_scanner
+		to_chat(owner, "<span class='notice'>Research analyzer is now [owner.research_scanner ? "active" : "deactivated"].</span>")
+		return 1
+
+/datum/action/item_action/toggle_research_scanner/Remove(mob/living/L)
+	if(owner)
+		owner.research_scanner = 0
+	..()
+
+/datum/action/item_action/toggle_research_scanner/ApplyIcon(obj/screen/movable/action_button/current_button)
+	if(button_icon && button_icon_state)
+		var/image/img = image(button_icon, current_button, "scan_mode")
+		current_button.overlays += img
+
+/datum/action/item_action/remove_badge
+	name = "Remove Holobadge"
+
+///prset for organ actions
+/datum/action/item_action/organ_action
+	check_flags = AB_CHECK_CONSCIOUS
+
+/datum/action/item_action/organ_action/IsAvailable()
+	var/obj/item/organ/internal/I = target
+	if(!I.owner)
+		return 0
+	return ..()
+
+/datum/action/item_action/organ_action/toggle
+/datum/action/item_action/organ_action/toggle/New(Target)
+	..()
+	name = "Toggle [target.name]"
+	button.name = name
+
+// for clothing accessories like holsters
+/datum/action/item_action/accessory
+	check_flags = AB_CHECK_RESTRAINED|AB_CHECK_STUNNED|AB_CHECK_LYING|AB_CHECK_CONSCIOUS
+
+/datum/action/item_action/accessory/IsAvailable()
+	. = ..()
+	if(!.)
+		return 0
+	if(target.loc == owner)
+		return 1
+	if(istype(target.loc, /obj/item/clothing/under) && target.loc.loc == owner)
+		return 1
+	return 0
+
+/datum/action/item_action/accessory/holster
+	name = "Holster"
+
+/datum/action/item_action/accessory/storage
+	name = "View Storage"
+
+
+//Preset for spells
+/datum/action/spell_action
+	check_flags = 0
+	background_icon_state = "bg_spell"
+
+/datum/action/spell_action/New(Target)
+	..()
+	var/obj/effect/proc_holder/spell/S = target
+	S.action = src
+	name = S.name
+	button_icon = S.action_icon
+	button_icon_state = S.action_icon_state
+	background_icon_state = S.action_background_icon_state
+	button.name = name
+
+/datum/action/spell_action/Destroy()
+	var/obj/effect/proc_holder/spell/S = target
+	S.action = null
+	return ..()
+
+/datum/action/spell_action/Trigger()
+	if(!..())
+		return 0
+	if(target)
+		var/obj/effect/proc_holder/spell = target
+		spell.Click()
+		return 1
+
+/datum/action/spell_action/IsAvailable()
```

Comment:
this won't work for latejoining mimes, apparently: https://github.com/tgstation/tgstation/pull/17973